### PR TITLE
Intro

### PR DESCRIPTION
**Pull request by lucica28**
Lab 1 for D summer school. Added quizes and updated exercises. Support for quizes is not yet implemented. Nested scopes and invariants were removed. I suggest we move nested scopes to lab 3 or 4 (with classes/structs) and we omit invariants.

### CODE REVIEW

File: intro/README.md [hunk 1]
```diff
@@ -0,0 +1,515 @@
+# Introduction to D
+
+
+## Syntax
+
+The D programming language uses a C-style syntax that ensures a smooth transition for programmers coming from a C\C++ background. With no further ado, let us take a random C program and see what it takes to compile it with the D compiler:
+
```
Comment (RazvanN7) [inline intro/README.md:7]:
```d
...cod
```

File: intro/README.md [hunk 2]
```diff
@@ -0,0 +1,629 @@
...
+The full D grammar can be found [here](https://dlang.org/spec/grammar.html).
+
+Now, using the above code snippet, let us delve into the D specific concepts.
+
+## Imports
+
+In D, **imports** represent the counterpart of the C **include** directive, however the are 2 fundamental differences:
+
+1. Imports may selectively specify which symbols are to be imported. For example, in the above code snippet, the full standard IO module of the standard library is imported, even though only the **printf** function is used. This results in a degradation in compile time since there is a larger symbol pool that needs to be examined when trying to resolve a symbol. In order to fix this, we can replace the blunt import with:
+        import std.stdio : printf;
```
Comment (RazvanN7) [inline intro/README.md:64]: use code block for this too.

File: intro/README.md [hunk 3]
```diff
@@ -0,0 +1,629 @@
...
+
+2. Imports may be used at any scope level.
+    ```d
+    int main()
```
Comment (RazvanN7) [inline intro/README.md:68]: this code is indented differently from the above snippets (it contains an extra tab)

File: intro/README.md [hunk 4]
```diff
@@ -0,0 +1,629 @@
...
+
+We will not insist on basic data types and pointers, as those are the same (or slightly modified versions) as the ones in C\C++. We will focus on arrays, associative arrays, classes, structs and unions. Delegates, functions and enums will be treated in a future lab.
+
+Note that in D all types have a default value. This means that there are no uninitialized variables.
+
+## Arrays
+
+The fundamental difference between a C array and a D array is that the former is represented by a simple pointer, while the latter is represented by a pointer and a size. This design decision was taken because of the numerous cases of C buffer overflow attacks that can be simply mitigated by automatically checking the array bounds. Let's take a simple example:
+
+```d
```
Comment (RazvanN7) [inline intro/README.md:103]:
```suggestion
```c
```

File: intro/README.md [hunk 5]
```diff
@@ -0,0 +1,629 @@
...
+void main()
+{
+    import std.stdio : writeln;
+
+    int[10] a;
+    auto b = a[0 .. $];
+    b = b ~ 5;
+
+    writeln(b.length);
+}
```
Comment (RazvanN7) [inline intro/README.md:152]: This quiz is too early. We have not talked about the `$` and `~` operators.

File: intro/README.md [hunk 6]
```diff
@@ -0,0 +1,629 @@
...
+         b = a[];                   // semantically equivalent to the one above
+         b = a[0 .. a.length]       // semantically equivalent to the one above
+         b[] = a[];                 // semantically equivalent to the one above
+         b[2 .. 4] = 4              // same as b[2] = 4, b[3] = 4
+         b[0 .. 4] = a[0 .. 4]      // error, a does not have 4 elements
+         a[0 .. 3] = b              // error, operands have different length
+    }
+    ```
+        
+    **.length** is a builtin array property. For an extensive list of array properties click [here](https://dlang.org/spec/arrays.html#array-properties).
```
Comment (RazvanN7) [inline intro/README.md:182]: Why is the indentation here different from the rest of the text?

File: intro/README.md [hunk 7]
```diff
@@ -0,0 +1,629 @@
...
+    void main()
+    {
+        int a = 5;
+        int b[10];
+        b[11] = 9;
+        printf("%d\n", a);
+    }
+```
+
+Compiling this code (without the stack protector activated) will lead to a buffer overflow in which the variable `a` will be modified and the print will output `9`. Compiling the same code with D (simply replace the include with an `import std.stdio : printf`) will issue a compile time error that states that 11 is beyond the size of the array. Aggregating the pointer of the array with its size facilitates a safer implementation of arrays.
```
Comment (RazvanN7) [inline intro/README.md:115]: One sentence per line please.

File: intro/README.md [hunk 8]
```diff
@@ -0,0 +1,629 @@
...
+
+    a.startWith2();
+    writeln(a);
+}
+```
+    
+- [2, 2]
+- [2, 2, 0, 0, 0]
+- [0, 0, 0, 0, 0]
+- [0, 0]
```
Comment (RazvanN7) [inline intro/README.md:271]: At this quiz you are also outputting the length, however the answers contain only the array.

File: intro/README.md [hunk 9]
```diff
@@ -0,0 +1,629 @@
...
+
+### Exercise
+
+1. C to D
```
Comment (RazvanN7) [inline intro/README.md:275]: I think that this exercise could be moved right after the import part, no?

File: intro/README.md [hunk 10]
```diff
@@ -0,0 +1,629 @@
...
+
+    ##### For 2 and 3 you can start your implementation using the skeleton inside **array-median** directory.
+
+
+2. Compute the median element of an unsorted integer array. For this, you will have to:
+    - Sort the array: implement any sorting algorithm you wish.
+    - Eliminate the duplicates: once the array is sorted, eliminating the duplicates is trivial.
+    - Check if the length of the array is **odd** or **even** and then return the median element.
+
+3. D has a standard library called [phobos](https://dlang.org/phobos/). Implement exercise 2 using functions from the [std.algorithm](https://dlang.org/phobos/std_algorithm.html) package. Use UFCS for an increase in expressiveness. Observe the increase in performance brought by using functions from the standard library
```
Comment (RazvanN7) [inline intro/README.md:288]: We have not talked about ufcs or functions up until now.

File: intro/README.md [hunk 11]
```diff
@@ -0,0 +1,629 @@
...
+- 0, 2
+
+### Exercise
+
+Navigate to the struct-class directory. You will find 2 files implementing the same program. One uses a struct as a data container, while the other uses a class.
+
+1. Compile both files and measure the run time of each program. How do you explain the differences?
+2. In both situations, print the value of the field a0 after the loop ends. How do you explain the differences? How we can change the code for structs so both versions produce the same output?
+
+## Functions
```
Comment (RazvanN7) [inline intro/README.md:431]: Maybe you could add the 3rd exercise from the first pool of exercises after this paragraph?

**Review by RazvanN7 — CHANGES_REQUESTED** [review]
This is a great start, but still the text of the lab needs to be updated to:

- use one sentence per line
- the exercises should be better dispersed between concept explanation
- ideally, include another couple of exercises.

**Comment by RazvanN7** [comment]
It's not necessarily a blocker for this PR, but we should definitely include some basic git information.

File: intro/README.md [hunk 12]
```diff
@@ -0,0 +1,683 @@
...
+    Auto can be also used to infer the type of a variable declaration.
+    
+### Exercise
+
+Navigate to the **voldemort** directory. Inspect the source file **voldermort.d**.
+The declaration of struct **Result** is declared as **private** (nobody has access to it, except the members of the current file).
+Move the declaration of the struct **Result** inside the **fun** function.
+Compile the code. Does it compile? Why? Fix the issue.
+    
+## Unittests
```
Comment (teodutu) [inline intro/README.md:523]: IMO, `unittests` are really important. Add an exercise here. Start from a demo where some `unittests` fail and have the students fix them, then add a few new ones.

File: intro/README.md [hunk 13]
```diff
@@ -0,0 +1,683 @@
...
+
+### Preconditions/Postconditions
+
+The pre contracts specify the preconditions before a statement is executed. The most typical use of this would be in validating the parameters to a function.
+The post contracts validate the result of the statement.
+The most typical use of this would be in validating the return value of a function and of any side effects it has.
+In D, pre contracts begin with in, and post contracts begin with out.
+They come at the end of the function signature and before the opening brace of the function body.
+
+```d
```
Comment (teodutu) [inline intro/README.md:568]: This snippet can be moved to a demo and extended to an exercise.

File: intro/README.md [hunk 14]
```diff
@@ -0,0 +1,683 @@
...
+For public or exported functions, the order of execution is:
+1. preconditions
+2. invariant
+3. function body
+4. invariant
+5. postconditions
+
+### Quiz
+
+What is the output of the following code?
```
Comment (teodutu) [inline intro/README.md:634]: In addition, "what is the output of this program" is not a very relevant quiz question in general. We don't want to teach students to be human compilers.

File: intro/README.md [hunk 15]
```diff
@@ -0,0 +1,683 @@
...
+- Runtime error at 1st call of fun()
+- Runtime error at 2nd call of fun()
+- hello world!
+-42
+- -42
+Hello World!
+
+### Exercise
+
+Navigate to the binary-search directory.
```
Comment (teodutu) [inline intro/README.md:669]:
```suggestion
Navigate to the `binary-search` directory.
```

File: intro/README.md [hunk 16]
```diff
@@ -0,0 +1,683 @@
...
+Inspect the source file binarySearch.d.
```
Comment (teodutu) [inline intro/README.md:670]:
```suggestion
Inspect the source file `binary_search.d.`
```

File: intro/README.md [hunk 17]
```diff
@@ -0,0 +1,683 @@
...
+As the name implies, a binarySearch algorithm is implemented on integers.
+Compile and run the file.
+
+- Write a unittest function that tests some corner cases. Are there any bugs in the algorithm implementation? If yes, fix them.
```
Comment (teodutu) [inline intro/README.md:674]:
```suggestion
- Write a unittest function that tests some corner cases.
Are there any bugs in the algorithm implementation?
If so, fix them.
```

File: intro/README.md [hunk 18]
```diff
@@ -0,0 +1,683 @@
...
+- The binarySearch function may be called with invalid data (for example: l = -1, r = -2). Write an in contract that halts the code execution in the case of invalid input (invalid values for l and r, the array is not sorted etc.)
```
Comment (teodutu) [inline intro/README.md:675]:
```suggestion
- The `binarySearch` function may be called with invalid data (for example: `l` = -1, `r` = -2).
Write an `in` contract that halts the code execution in the case of invalid input (invalid values for l and `r`, the array is not sorted etc.)
```

File: intro/README.md [hunk 19]
```diff
@@ -0,0 +1,683 @@
...
+- Rewrite the binarySearch algorithm to make use of slices.
+
+### Exercise
+
+Navigate to array-median element. Follow the directions in the skeleton in order to implement our very own string : int associative array.
```
Comment (teodutu) [inline intro/README.md:680]:
```suggestion
Navigate to the `array-median` folder.
Follow the directions in the skeleton in order to implement our very own `string : int` associative array.
```

File: intro/array-median/arrayMedian.d [hunk 20]
```diff
@@ -0,0 +1,46 @@
+/* Compile and run: dmd -main -unittest -run arrayMedian.d */
+
```
Comment (teodutu) [inline intro/array-median/arrayMedian.d:2]:
```suggestion
```

File: intro/array-median/arrayMedian.d [hunk 21]
```diff
@@ -0,0 +1,46 @@
...
+    auto r = benchmark!(callMedianElem, callMedianElemPhobos)(1);
+    Duration noPhobosRes = r[0];
+    Duration phobosRes   = r[1];
+
+    writeln("Duration for 100000 without phobos:");
+    writeln(noPhobosRes);
+    writeln("Duration for 100000 with phobos:");
+    writeln(phobosRes);
+}
```
Comment (teodutu) [inline intro/array-median/arrayMedian.d:46]:
```suggestion
}

```

File: intro/majority-element/majorityElement.d [hunk 22]
```diff
@@ -0,0 +1,17 @@
+/* Compile and run: dmd -main -unittest -run majorityElement.d */
+
```
Comment (teodutu) [inline intro/majority-element/majorityElement.d:2]:
```suggestion
```

File: intro/my-associative-array/Makefile [hunk 23]
```diff
@@ -0,0 +1,5 @@
+DC = dmd
+DFLAGS = -main -unittest -run
+
+all:
+	$(DC) $(DFLAGS) myaa.d
```
Comment (teodutu) [inline intro/my-associative-array/Makefile:5]:
```suggestion
	$(DC) $(DFLAGS) myaa.d

```

File: intro/my-associative-array/myaa.d [hunk 24]
```diff
@@ -0,0 +1,83 @@
...
+
+    MyAssociativeArray otherAa = aa.clone();
+    assert(otherAa.get("hello") == 3);
+    otherAa.add("D", 9);
+    assert(otherAa.get("D") == 9);
+    assert(aa.get("D") == 0);
+    otherAa.add("Dss", 10);
+    assert(otherAa.keys().sort() == ["hello", "D", "Dss"].sort());
+    writeln("Congratulations, all tests passed!");
+}
```
Comment (teodutu) [inline intro/my-associative-array/myaa.d:83]:
```suggestion
}

```

File: intro/struct-class/Makefile [hunk 25]
```diff
@@ -0,0 +1,10 @@
+DC = dmd
+
+struct:
+	$(DC) struct.d
+
+class:
+	$(DC) class.d
+
+clean:
+	rm *.o class struct
```
Comment (teodutu) [inline intro/struct-class/Makefile:10]:
```suggestion
	rm *.o class struct

```

File: intro/struct-class/struct.d [hunk 26]
```diff
@@ -0,0 +1,30 @@
...
+void get_a0(BigStruct a)
+{
+    var = a.a0;
+    a.a0 = 42;
+}
+
+void main()
+{
+    BigStruct s;
+    for(int i = 0; i < 1000000; i++)
```
Comment (teodutu) [inline intro/struct-class/struct.d:28]:
```suggestion
    for (int i = 0; i < 1000000; i++)
```

File: intro/struct-class/class.d [hunk 27]
```diff
@@ -0,0 +1,30 @@
...
+void get_a0(BigClass a)
+{
+    var = a.a0;
+    a.a0 = 42;
+}
+
+void main()
+{
+    BigClass s = new BigClass();
+    for(int i = 0; i < 100000; i++)
```
Comment (teodutu) [inline intro/struct-class/class.d:28]:
```suggestion
    for (int i = 0; i < 100000; i++)
```

File: intro/voldemort/Makefile [hunk 28]
```diff
@@ -0,0 +1,7 @@
+DC = dmd
+
+all:
+	$(DC) voldemort.d
+
+clean:
+	rm *.o voldemort
```
Comment (teodutu) [inline intro/voldemort/Makefile:7]:
```suggestion
	rm *.o voldemort

```

**Review by teodutu — CHANGES_REQUESTED** [review]
Apart from the inline suggestions, here are a few higher-level ones:
- Change the quiz questions from "what does this code do?" to "why does(n't) this happen".
- Change the names of `.d` files to `snake_case`.
- Change exercise sections `### Exercise` to `### Practice`. It sounds less formal and more hands-on.
- Add the quizzes to a separate file in a new folder called `quiz`. Use the same format as here: https://github.com/Dlang-UPB/D-Summer-School/pull/12.
- Change numbered lists from using `1. 2. 3. ...` to `1. 1. 1. ...`.
- Add a `clean` rule to all Makefiles that removes the executable file and the object file. Use `rm -f` in the `clean` rule.
- Make both `all` and `clone` rules `.PHONY` in each Makefile.
- Create a single makefile in `intro/` and `include` it in all other Makefiles. Pretty much like this:
    - https://github.com/Dlang-UPB/D-Summer-School/blob/550b2367af78b4ee00c5751c6f2e5e786cf79de6/advanced-metaprogramming/Makefile
    - https://github.com/Dlang-UPB/D-Summer-School/blob/550b2367af78b4ee00c5751c6f2e5e786cf79de6/advanced-metaprogramming/demo/compile-time-reflection/Makefile
- Add a `.gitignore` file for each exercise that ignores the executable file.
- I'd turn some of the code snippets into demos. Ideally, some demos should not work intentionally and their corresponding exercise (or part of it) should be to make them work.

File: intro/README.md [hunk 29]
```diff
@@ -0,0 +1,683 @@
...
+
+    return 0;
+}
+```
+
+### Exercise
+
+C to D
+Go to this [link](https://github.com/TheAlgorithms/C/tree/master/searching). You will find a series of searching algorithms, each implemented in C in its own file.
+- Choose one algorithm and port it to D with minimum modifications.
```
[the remark this answers is not in the document]
Comment (lucica28) [reply intro/README.md:96]: maybe we can get rid of this?

File: intro/Makefile [hunk 30]
```diff
@@ -0,0 +1,12 @@
...
+
+all: $(BINARY)
+
+$(BINARY): $(BINARY).d
+	$(DMD) $(DFLAGS) $(BINARY).d
+
+clean:
+	rm -f $(BINARY) $(BINARY).o
+
+.PHONY: all clean
```
Comment (teodutu) [inline intro/Makefile:12]:
```suggestion
.PHONY: all clean

```

File: intro/README.md (hunk 29 above)
[the remark this answers is not in the document]
Comment (teodutu) [reply intro/README.md:96]: Not sure. I think it's a good first exercise to just port something from C to D an see how easy it is (supposedly).

File: intro/README.md [hunk 31]
```diff
@@ -0,0 +1,561 @@
...
+
+## Classes
+
+In D, classes are very similar with java classes:
+
+1. Classes can implement any number of interfaces
+2. Classes can inherit a single class
+3. Class objects are instantiated by reference only
+4. super has the same meaning as in Java
+5. overriding rules are very similar
```
Comment (teodutu) [inline intro/README.md:346]:
```suggestion
1. Classes can inherit a single class
1. Class objects are instantiated by reference only
1. super has the same meaning as in Java
1. overriding rules are very similar
```

File: intro/README.md [hunk 32]
```diff
@@ -0,0 +1,561 @@
...
+
+The fundamental difference between **structs** and **classes** is that the former are **value** types, while the latter are **reference** types.
+This means that whenever a struct is passed as an argument to an lvalue function parameter, the function will operate on a copy of the object.
+When a class is passed as an argument to an lvalue function parameter, the function will receive a reference to the object.
```
Comment (teodutu) [inline intro/README.md:350]: Does it make sense to explain what a reference is? I think 1 / 2 phrases that it's actually a pointer blah blah wouldn't hurt.

File: intro/README.md [hunk 33]
```diff
@@ -0,0 +1,561 @@
...
+
+Both **structs** and **classes** will be covered more in depth in a further lab.
+
+Navigate to the **demo/struct-class** directory.
```
Comment (teodutu) [inline intro/README.md:354]:
```suggestion
Navigate to the `demo/struct-class` directory.
```

File: intro/README.md [hunk 34]
```diff
@@ -0,0 +1,561 @@
...
+    {
+        int b = fun();
+        auto c = fun();   // works for variables also
+    }
+    ```
+        
+    Auto functions have their return type inferred based on the type of the return statements.
+    Auto can be also used to infer the type of a variable declaration.
+    
+Navigate to **demo/voldemort** directory. Inspect and run the code inside **voldermort.d**.
```
Comment (teodutu) [inline intro/README.md:424]:
```suggestion
Navigate to `demo/voldemort` directory. Inspect and run the code inside `voldermort.d`.
```

File: intro/README.md [hunk 35]
```diff
@@ -0,0 +1,561 @@
...
+
+Now navigate to **intro/voldemort** directory.
```
Comment (teodutu) [inline intro/README.md:426]:
```suggestion
Now navigate to `intro/voldemort` directory. 
```

File: intro/README.md [hunk 36]
```diff
@@ -0,0 +1,561 @@
...
+    }
+    ```
+        
+    Auto functions have their return type inferred based on the type of the return statements.
+    Auto can be also used to infer the type of a variable declaration.
+    
+Navigate to **demo/voldemort** directory. Inspect and run the code inside **voldermort.d**.
+
+Now navigate to **intro/voldemort** directory. 
+**Result** is a struct declared inside **fun's** scope.
```
Comment (teodutu) [inline intro/README.md:427]:
```suggestion
`Result` is a struct declared inside `fun`'s scope.
```

File: intro/README.md [hunk 37]
```diff
@@ -0,0 +1,561 @@
...
+However we want to be able to use it in **main** as well.
```
Comment (teodutu) [inline intro/README.md:428]:
```suggestion
However, we want to be able to use it in `main` as well.
```

File: intro/README.md [hunk 38]
```diff
@@ -0,0 +1,561 @@
...
+Inspect the file. Does it compile? Why? Fix the issue.
```
Comment (teodutu) [inline intro/README.md:429]:
```suggestion
Inspect the file.
Does it compile?
Why?
Fix the issue.
```

File: intro/README.md [hunk 39]
```diff
@@ -0,0 +1,561 @@
...
+    unittest
+    {
+        Sum sum;
+        assert(sum.add(3,4) == 7);
+        assert(sum.add(-2,0) == -2);
+    }
+}
+```
+
+Navigate to the **demo/binary-search** directory.
```
Comment (teodutu) [inline intro/README.md:464]:
```suggestion
Navigate to the `demo/binary-search` directory.
```

File: intro/README.md [hunk 40]
```diff
@@ -0,0 +1,561 @@
...
+Inspect the source file binarySearch.d.
```
Comment (teodutu) [inline intro/README.md:465]:
```suggestion
Inspect the source file `binarySearch.d`.
```

File: intro/README.md [hunk 41]
```diff
@@ -0,0 +1,561 @@
...
+As the name implies, a binarySearch algorithm is implemented on integers.
+Compile and run the file.
+
+- Write a unittest function that tests some corner cases. Are there any bugs in the algorithm implementation? If yes, fix them.
```
Comment (teodutu) [inline intro/README.md:469]:
```suggestion
- Write a unittest function that tests some corner cases.
Are there any bugs in the algorithm implementation?
If so, fix them.
```

File: intro/README.md [hunk 42]
```diff
@@ -0,0 +1,561 @@
...
+- Rewrite the binarySearch algorithm to make use of slices.
+   
+## Contract Programming
+
+Contracts are a breakthrough technique to reduce the programming effort for large projects. Contracts are the concept of preconditions, postconditions and errors.
```
Comment (teodutu) [inline intro/README.md:474]:
```suggestion
Contracts are a breakthrough technique to reduce the programming effort for large projects.
Contracts are the concept of preconditions, postconditions and errors.
```

File: intro/README.md [hunk 43]
```diff
@@ -0,0 +1,561 @@
...
+
+### Preconditions/Postconditions
+
+The pre contracts specify the preconditions before a statement is executed. The most typical use of this would be in validating the parameters to a function.
```
Comment (teodutu) [inline intro/README.md:478]:
```suggestion
The pre contracts specify the preconditions before a statement is executed.
The most typical use of this would be in validating the parameters to a function.
```

File: intro/README.md [hunk 44]
```diff
@@ -0,0 +1,561 @@
...
+    }
+}
+```
+
+For public or exported functions, the order of execution is:
+1. preconditions
+2. invariant
+3. function body
+4. invariant
+5. postconditions
```
Comment (teodutu) [inline intro/README.md:555]:
```suggestion
1. invariant
1. function body
1. invariant
1. postconditions
```

File: intro/README.md [hunk 45]
```diff
@@ -0,0 +1,561 @@
...
+
+Navigate to **my-associative-array** directory.
```
Comment (teodutu) [inline intro/README.md:557]:
```suggestion
Navigate to the `my-associative-array` directory.
```

File: intro/README.md [hunk 46]
```diff
@@ -0,0 +1,561 @@
...
+Follow the directions in the skeleton in order to implement our very own string : int associative array.
```
Comment (teodutu) [inline intro/README.md:558]:
```suggestion
Follow the directions in the skeleton in order to implement our very own `string : int` associative array.
```

**Review by teodutu — CHANGES_REQUESTED** [review]
There are still some unresolved inline comments. They are mostly cosmetic changes which shouldn't take you very long, such as changing the names of the `.d` files to `camel_case` instead of `snakeCase`. I found some other cosmetic inconsistencies when looking at the lab again.

In addition, leave an empty line at the end of each file. For example, the new makefiles look alright, but they don't end in a new line. This is how POSIX defines a line [1]. It must end with a `\n`.

[1] https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap03.html#tag_03_206

File: intro/README.md (hunk 29 above)
[the remark this answers is not in the document]
Comment (RazvanN7) [reply intro/README.md:96]: Yes, the exercise is good. I like the fact that they can choose. I suggest we leave this as is.

File: intro/README.md [hunk 47]
```diff
@@ -0,0 +1,561 @@
...
+
+    return 0;
+}
+```
+
+### C to D
+
+Go to this [link](https://github.com/TheAlgorithms/C/tree/master/searching). You will find a series of searching algorithms, each implemented in C in its own file.
+- Choose one algorithm and port it to D with minimum modifications.
+- Update the code using D specific features to improve the code (fewer lines of code, increase in expressiveness etc.).
```
Comment (RazvanN7) [inline intro/README.md:97]: This should be done later. Right now we haven't shown anything about how powerful D is.

File: intro/README.md [hunk 48]
```diff
@@ -0,0 +1,561 @@
...
+        a[0 .. 3] = b              // error, operands have different length
+    }
+    ```
+        
+    **.length** is a builtin array property.
+    For an extensive list of array properties click [here](https://dlang.org/spec/arrays.html#array-properties).
+
+    Now let's see how arrays work with functions.
+    The next quiz can be tricky but it will help us better understand how arrays work.
+    If it help, think what would happed in a similar situation in C.
```
Comment (RazvanN7) [inline intro/README.md:186]:
```suggestion
    If it helps, think what would happed in a similar situation in C.
```

File: intro/README.md [hunk 49]
```diff
@@ -0,0 +1,561 @@
...
+    
+    [Quiz](./quiz/ref-vs-value.md)
```
Comment (RazvanN7) [inline intro/README.md:188]: This quiz requires knowledge about structs. Up until this point we haven't explained what structs are.

Comment (RazvanN7) [reply intro/README.md:188]: I suggest putting this quiz after we talk about structs and after we explain the fact that an array in D is seen as a struct that contains the pointer and a size.

File: intro/README.md [hunk 50]
```diff
@@ -0,0 +1,561 @@
...
+     
+        a = b ~ c;       // a will be [1, 2, 3, 4, 5]
+        a = b;           // a refers to b
+        a = b ~ c[0..0]; // a refers to a copy of b
+        a ~= c;          // equivalent to a = a ~ c;
+    }
+    ```
+       
+    Concatenation always creates a copy of its operands, even if one of the operands is a 0-length array.
+    The operator `~=` does not always create a copy.
```
Comment (RazvanN7) [inline intro/README.md:207]: This is confusing. You should provide more information about this. When does `~=` create a copy and when does it not?

File: intro/README.md [hunk 51]
```diff
@@ -0,0 +1,561 @@
...
+    For more information click [here](https://dlang.org/spec/arrays.html#resize).
+
+    Navigate to **demo/slice** directory.
+    Inspect and run the file **slice.d**.
+    What happened here? Why?
+
+    [Quiz](./quiz/slice.md)
```
Comment (RazvanN7) [inline intro/README.md:214]: You could add here the exercise to improve the initial example by using slices instead of the for loop.

File: intro/README.md [hunk 52]
```diff
@@ -0,0 +1,683 @@
...
+        b[] = a[] + 4;         // b = [5, 6, 7]
+        c[] = (a[] + 1) * b[]  // c = [10, 18, 28]
+    }
+    ```
+    
+    Many array operations, also known as vector operations, can be expressed at a high level rather than as a loop.
+    A vector operation is indicated by the slice operator appearing as the left-hand side of an assignment or an op-assignment expression.
+    The right-hand side can be an expression consisting either of an array slice of the same length and type as the left-hand side or a scalar expression of the element type of the left-hand side, in any combination.
+
+    Using the concepts of slicing and concatenation, we can modify the original example (that does the removal of an element from the array) so that the **for** loop is no longer necessary:
```
[the remark this answers is not in the document]
Comment (RazvanN7) [reply intro/README.md:234]: You could maybe ask them to do this.

[the remark this answers is not in the document]
Comment (RazvanN7) [reply intro/README.md:234]: On a second thought. It would be better to leave this example and then ask them to improve the code from the first exercise by using slices.

File: intro/README.md [hunk 53]
```diff
@@ -0,0 +1,561 @@
...
+    ```
+    
+    The “in” expression yields a pointer to the value if the key is in the associative array, or null if not.
+
+    For more advanced operations on AAs check this [link](https://dlang.org/spec/hash-map.html#construction_assignment_entries).
+    For an exhaustive list of the AA properties check this [link](https://dlang.org/spec/hash-map.html#properties).
+
+
+Find the [majority element](https://leetcode.com/problems/majority-element/) in a string array using builtin associative arrays.
+You can start your implementation using the skeleton inside **majority-element** directory
```
Comment (RazvanN7) [inline intro/README.md:319]:
```suggestion
You can start your implementation using the skeleton inside **majority-element** directory.
```

File: intro/README.md [hunk 54]
```diff
@@ -0,0 +1,561 @@
...
+Both **structs** and **classes** will be covered more in depth in a further lab.
+
+Navigate to the **demo/struct-class** directory.
+You will find 2 files implementing the same program.
+One uses a struct as a data container, while the other uses a class.
+
+- Compile both files and measure the run time of each program. How do you explain the differences?
+- How do you explain the output differences? How we can change the code for structs so both versions produce the same output?
+
+[Quiz](./quiz/struct-class.md)
```
Comment (RazvanN7) [inline intro/README.md:361]: I think that the quiz ref-vs-value is much more suited here after they understand what a value type is. You can have any number of consecutive quizzes.

File: intro/README.md [hunk 55]
```diff
@@ -0,0 +1,561 @@
...
+    ```d
+    import std.algorithm : group;
+    import std.range : chain, dropOne, front, retro;
+    [1, 2].chain([3, 4]).retro; // 4, 3, 2, 1
+    [1, 1, 2, 2, 2].group.dropOne.front; // (2, 3)
+     
+    front(dropOne(group([1, 1, 2, 2, 2])));
+    ```
+
+D has a standard library called [phobos](https://dlang.org/phobos/).
```
Comment (RazvanN7) [inline intro/README.md:378]: This paragraph is very abrupt. You could maybe put the exercise first and then instruct them to search for functions  in the standard library. Something along the lines of: "Find the median [...]. Use any number of functions from the [standard library](link). Use UFCS to make the code more expressive."

File: intro/README.md [hunk 56]
```diff
@@ -0,0 +1,561 @@
...
+Find the **median element** using functions from the [std.algorithm](https://dlang.org/phobos/std_algorithm.html) package.
+Use UFCS for an increase in expressiveness.
+Observe the increase in performance achieved by using functions from the standard library
```
Comment (RazvanN7) [inline intro/README.md:381]: Is this a task? What are they supposed to do? Time the program?

File: intro/README.md [hunk 57]
```diff
@@ -0,0 +1,561 @@
...
+
+2. Overloading:
+
+    ```d
+    void fun(int a) {}
+    void fun(int a, int b) {}
+    ```
+    
+    Overloading is done based on the **type of parameters**, not on the **return type**.
```
Comment (RazvanN7) [inline intro/README.md:390]: This section needs to be expanded. You should explain a bit what overloading is and also the example needs to contain some calls to these functions.

File: intro/README.md (hunk 56 above)
Comment (lucica28) [reply intro/README.md:381]: Not really, just run the program, in the skeleton they should complete 2 functions, 1 using the standard library, and one without. Then, inside the unittest there is an example and they are timed.

**Review by RazvanN7 — APPROVED** [review]
Looks a lot better, but please address the review points.

File: intro/README.md (hunk 51 above)
Comment (lucica28) [reply intro/README.md:214]: If you are talking about the exercise with binary search where they have to fix a bug, I feel like the part with the unittest should be first, because chances are they will fix the bug without even realizing, and I think the main purpose of that exercise is to show how useful unittests can be

Comment (RazvanN7) [reply intro/README.md:214]: No, I was referring to the initial example where an element is deleted from an array.

File: intro/README.md [hunk 58]
```diff
@@ -0,0 +1,618 @@
...
+
+    For more advanced operations on AAs check this [link](https://dlang.org/spec/hash-map.html#construction_assignment_entries).
+    For an exhaustive list of the AA properties check this [link](https://dlang.org/spec/hash-map.html#properties).
+
+## Practice
+
+Find the [majority element](https://leetcode.com/problems/majority-element/) in a string array using builtin associative arrays.
+You can start your implementation using the skeleton inside `majority-element` directory.
+
+### C to D
```
Comment (RazvanN7) [inline intro/README.md:329]: Shouldn't this also be tagged as "Practice" ?

**Comment by RazvanN7** [comment]
This seems to be in good shape now. We can fix other small things as we identify them.
